src/javascript/crypto/e2e/vrf/extra25519.js

```diff
@@ -0,0 +1,185 @@
+/**
+ * @license
+ * Copyright 2015 Yahoo Inc. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Helper functions for the Yahoo E2E vrf.verify() implementation
+ * This is ported from https://github.com/yahoo/coname/blob/master/vrf/vrf.go
+ */
+
+goog.provide('e2e.vrf.extra25519');
+
+goog.require('goog.asserts');
+goog.require('goog.crypt');
+goog.require('e2e.BigNum');
+goog.require('e2e.ecc.constant');
+goog.require('e2e.ecc.DomainParam');
+goog.require('e2e.ecc.PrimeCurve');
+
+goog.require('e2e.vrf.sha3Shake256');
+
+e2e.vrf.ed25519 = e2e.ecc.DomainParam.fromCurve(e2e.ecc.PrimeCurve.ED_25519);
+e2e.vrf.curve25519 = e2e.ecc.DomainParam.fromCurve(e2e.ecc.PrimeCurve.CURVE_25519);
+
+
+/**
+ * Unmarshals an elliptic curve point, and validates that it is an Ed25519 public key
+ * Refer to Definition 1 in https://www.iacr.org/archive/pkc2003/25670211/25670211.pdf
+ * 
+ * @param {!e2e.ByteArray} sBytes A byte array of length 32
+ * @return {!e2e.ecc.point.Ed25519}
+ */
+e2e.vrf.extra25519.fromBytesBaseGroup = function(sBytes) {
+
+	var P = e2e.vrf.ed25519.curve.pointFromByteArray(sBytes);
+
+	goog.asserts.assert(!P.isInfinity(),
+	// goog.asserts.assert(!P.isIdentity(),
+        'P != infinity');
+
+	goog.asserts.assert(P.isOnCurve(),
+        'P must satisfy the defining equation of the elliptic curve E');
+
+	// TODO: confirm https://github.com/google/end-to-end/issues/340
```

Comment:
Hmm. I think isIdentity is the way to go, and I am not sure what meaning `this.INFINITY = new e2e.ecc.point.Ed25519(this,  this.ZERO, this.ZERO, this.ZERO, this.ZERO)` has.

The reason we want this check is that we want all points used in the vrf computation to be in the prime-order subgroup of the non-prime-order elliptic curve. Multiplying by order of the subgroup should give the identity element (of both the subgroup and the curve) in that case. This reasoning does not involve the geometric ("at infinity") interpretation at all.

http://iml.univ-mrs.fr/ati/crypto_puces/2011/slides/Mardi/rolland.pdf and https://christianepeters.files.wordpress.com/2012/10/20080620-rennes.pdf say that projective(0,1,0) (the identity) is a point at infinity. https://hyperelliptic.org/EFD/oldefd/edwards.html and https://en.wikipedia.org/wiki/Edwards_curve#Projective_homogeneous_coordinates confirm that that is the neutral element (with no mention of infinity).

Traditional NIST curves only have one point at infinity and this is the identity. In edwards curves, we cannot assume that. Which still does not explain the value of `INFINITY` this library defines for ed25519, but oh well. I also find it interesting that the ed25519 code comments mention jacobian coordinates but reference the EFD entry for addition law in the (extended) projective coordinates.